cli/cli
# Add `pr update-branch` command

This PR adds the new `pr update-branch` command, to update a PR branch with the latest changes of its base. Under the hood, we use the [`updatePullRequestBranch`](https://docs.github.com/en/graphql/reference/mutations#updatepullrequestbranch) mutation.


Fixes #8426

## Details

The command doc is:

```text
Update a pull request branch with latest changes of the base branch.

Without an argument, the pull request that belongs to the current branch is selected.

The default behavior is to update with a merge commit (i.e., merging the base branch
into the the PR's branch). To reconcile the changes with rebasing on top of the base
branch, the `--rebase` option should be provided.


USAGE
  gh pr update-branch [<number> | <url> | <branch>] [flags]

FLAGS
  --rebase   Update PR branch by rebasing on top of latest base branch

INHERITED FLAGS
      --help                     Show help for command
  -R, --repo [HOST/]OWNER/REPO   Select another repository using the [HOST/]OWNER/REPO format

EXAMPLES
  $ gh pr update-branch 23
  $ gh pr update-branch 23 --rebase
  $ gh pr update-branch 23 --repo owner/repo

LEARN MORE
  Use `gh <command> <subcommand> --help` for more information about a command.
  Read the manual at https://cli.github.com/manual
```

## QA

I was able to QA the changes using the script below. This should be run at the repo's root directory, with the binary at `bin/gh`. Note that the script will nevertheless clean up on exit.

```sh
#!/usr/bin/env sh

set -e

_tmp=/tmp
_repo=gh-some-repo
_gh=$(pwd)/bin/gh
_pwd=$(pwd)

cleanup () {
    cd $_pwd
    ARG=$?
    rm -rf "$_tmp/$_repo"
    gh repo delete --yes $_repo
    exit $ARG
} 
trap cleanup EXIT

cd $_tmp
$_gh repo create --private --add-readme $_repo
$_gh repo clone $_repo
cd $_repo
git checkout -b pr-branch
echo "updated in pr branch" > README.md
git commit -am 'change in pr'
git push -u origin pr-branch
$_gh pr create --fill --title "test" --body ''
git checkout main
touch NEW.md
echo "new file on main" > NEW.md
git add NEW.md
git commit -am 'change on main'
git push -u origin main
git checkout pr-branch
$_gh pr update
```
